tiledb/api/src/query/read/aggregate.rs

```diff
@@ -0,0 +1,665 @@
+            let step_result = reader.step()?;
+            let enum_result = match step_result {
+                ReadStepOutput::Final((return_val, base_q)) => {
+                    ReadStepOutput::Final((
+                        AggregateResultHandle::from(return_val),
+                        base_q,
+                    ))
+                }
+                _ => unreachable!("Aggregate enum step function"),
+            };
+            Ok(enum_result)
+        })
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum AggregateType {
+    Count,
+    Max(String),
+    Mean(String),
+    Min(String),
+    NullCount(String),
+    Sum(String),
+}
+
+impl Display for AggregateType {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        <Self as Debug>::fmt(self, f)
+    }
+}
+
+fn get_datatype_from_attr(
+    schema: &Schema,
+    attr_name: &String,
+) -> TileDBResult<Datatype> {
+    let attr = schema.field(attr_name)?;
+    let attr_type = attr.datatype()?;
+    Ok(attr_type)
+}
+
+fn aggregate_type(
+    agg_type: &AggregateType,
+    schema: &Schema,
+) -> TileDBResult<Datatype> {
+    match agg_type {
+        AggregateType::Count | AggregateType::NullCount(_) => {
+            Ok(Datatype::UInt64)
+        }
+        AggregateType::Mean(_) => Ok(Datatype::Float64),
+        AggregateType::Sum(attr_name) => {
+            let attr_type = get_datatype_from_attr(schema, attr_name)?;
+            if attr_type == Datatype::Int8
+                || attr_type == Datatype::Int16
+                || attr_type == Datatype::Int32
+                || attr_type == Datatype::Int64
+            {
+                Ok(Datatype::Int64)
+            } else if attr_type == Datatype::UInt8
+                || attr_type == Datatype::UInt16
+                || attr_type == Datatype::UInt32
+                || attr_type == Datatype::UInt64
+            {
+                Ok(Datatype::UInt64)
+            } else if attr_type == Datatype::Float32
+                || attr_type == Datatype::Float64
+            {
+                Ok(Datatype::Float64)
+            } else {
+                Err(TileDBError::InvalidArgument(anyhow!(
+                    "Invalid attribute type."
+                )))
+            }
+        }
+        AggregateType::Min(attr_name) | AggregateType::Max(attr_name) => {
+            let attr_type = get_datatype_from_attr(schema, attr_name)?;
+            Ok(attr_type)
+        }
+    }
+}
+
+/// Trait for query types which can have an aggregate channel placed on top of them.
+pub trait AggregateBuilderTrait: QueryBuilder {
+    fn apply_typed_aggregate<T: 'static>(
+        self,
+        agg_type: AggregateType,
+    ) -> TileDBResult<AggregateTypedBuilder<Self, T>> {
+        let expected_type =
+            aggregate_type(&agg_type, &self.base().array().schema()?)?;
+        if !expected_type.is_compatible_type::<T>() {
+            return Err(TileDBError::InvalidArgument(anyhow!(
```

Comment:
This should be `Error::TypeMismatch` instead of InvalidArgument

Reply:
This takes a string so I just used typeID. Is there a better way of converting a type into a string?